prow/git/v2/remote_test.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"errors"
+	"net/url"
+	"reflect"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/util/diff"
+)
+
+func TestSimpleAuthResolver(t *testing.T) {
+	var testCases = []struct {
+		name        string
+		remote      func() (*url.URL, error)
+		username    func() (string, error)
+		token       func() []byte
+		expected    string
+		expectedErr bool
+	}{
+		{
+			name: "happy case works",
+			remote: func() (*url.URL, error) {
+				return &url.URL{
+					Scheme: "https",
+					Host:   "github.com",
+					Path:   "org/repo",
+				}, nil
+			},
+			username: func() (string, error) {
+				return "gitUser", nil
+			},
+			token: func() []byte {
+				return []byte("pass")
+			},
+			expected:    "https://gitUser:pass@github.com/org/repo",
+			expectedErr: false,
+		},
+		{
+			name: "failure to resolve remote URL creates error",
+			remote: func() (*url.URL, error) {
+				return nil, errors.New("oops")
+			},
+			expectedErr: true,
+		},
+		{
+			name: "failure to get username creates error",
+			remote: func() (*url.URL, error) {
+				return nil, nil
+			},
+			username: func() (string, error) {
+				return "", errors.New("oops")
+			},
+			expectedErr: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			actual, actualErr := simpleAuthResolver(testCase.remote, testCase.username, testCase.token)()
+			if testCase.expectedErr && actualErr == nil {
+				t.Errorf("%s: expected an error but got none", testCase.name)
+			}
+			if !testCase.expectedErr && actualErr != nil {
+				t.Errorf("%s: expected no error but got one: %v", testCase.name, actualErr)
+			}
+			if actual != testCase.expected {
+				t.Errorf("%s: got incorrect remote URL: %v", testCase.name, diff.StringDiff(actual, testCase.expected))
+			}
+		})
+	}
+}
+
+type stringWithError struct {
+	str string
+	err error
+}
+
+// usernameVendor returns a func that vends logins and errors from the provided slices, in order
+func usernameVendor(logins []stringWithError) func() (string, error) {
```

Comment:
I think the vendors have to be threadsafe, as we only lock per-repo/when accessing the repo lock map, but we could be accessing different repos in parallel, which would make this race.

Also, but this may be me not being a native speaker, I would never have guessed from the name what these do.

Reply:
This can only be racy if the function returned by this method is used in more than one place -- if multiple callers end up fielding it. This is test code and it's threadsafe in the way it's implemented (as one vendor is used per mock).